### ISSUE TITLE
Python mind agents can't call the methods of atomspace.

### ISSUE THREAD

**Issue by kim135797531** [opencog/opencog#1614]
I tried to create python mind agents, but I got the following error if I try to access any methods of atomspace.

Execute following code:
```python
class ConceptualBlendingAgent(opencog.cogserver.MindAgent):
    def run(self, atomspace):
        atomspace.add(types.Node, "test")
```

makes following error:
```
terminate called after throwing an instance of 'opencog::RuntimeException'
  what():  PyMindAgent triggered runtime exception.Traceback (most recent call last):
  File "agent_finder.pyx", line 145, in opencog.agent_finder.run_agent (/home/kdm/opencog/build/opencog/cython/opencog/agent_finder.cpp:2862)
  File "/home/kdm/opencog/opencog/python/blending_new/agents.py", line 10, in run
    atomspace.add(types.Node, "test")
AttributeError: 'opencog.agent_finder.AtomSpace' object has no attribute 'add'

[2015-05-13 04:53:23:660] [ERROR] Caught signal 6 (중지됨) on thread 140139092150336
	Stack Trace:
	2: Logger.h:175	  opencog::Logger::Base::~Base()
	3: CogServerMain.cc:79	_Z7sighandi()
	4: ??:0	killpg()
	5: raise.c:56	__GI_raise()
	6: abort.c:91	__GI_abort()
	7: ??:0	  __gnu_cxx::__verbose_terminate_handler()
	8: ??:0	_ZSt17rethrow_exceptionNSt15__exception_ptr13exception_ptrE()
	9: ??:0	_ZSt9terminatev()
	10: ??:0	__cxa_throw()
	11: PyMindAgent.cc:70	  opencog::PyMindAgent::run()
	12: BuiltinRequestsModule.cc:271	  opencog::BuiltinRequestsModule::do_stepAgents(opencog::Request*, std::list<std::string, std::allocator<std::string> >)
	13: BuiltinRequestsModule.h:111	  opencog::BuiltinRequestsModule::do_stepAgentsRequest::execute()
	14: CogServer.cc:297	  opencog::CogServer::processRequests()
	15: CogServer.cc:237	  opencog::CogServer::runLoopStep()
	16: CogServer.cc:200	  opencog::CogServer::serverLoop()
	17: CogServerMain.cc:211	main()
	18: libc-start.c:321	__libc_start_main()
	19: ??:0	_start()
```

Trying to find cause of this problem, I found some strange thing.

Result of ```print atomspace``` in python agent:
```<opencog.agent_finder.AtomSpace object at 0x7f41d1981370>```

Result of ```print AtomSpace()``` in cogserver python shell:
```<opencog.atomspace.AtomSpace object at 0x7f41d1981370>```

They have different type although they are located in same address. Due to this, list of methods are different:

Result of ```print dir(atomspace)``` in python agent:
```['__class__', '__delattr__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__']```

Result of ```print dir(AtomSpace())``` in cogserver python shell:
```['__class__', '__contains__', '__delattr__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'add', 'add_link', 'add_node', 'clear', 'get_atom_string', 'get_atoms_by_av', 'get_atoms_by_name', 'get_atoms_by_target_atom', 'get_atoms_by_type', 'get_atoms_in_attentional_focus', 'get_av', 'get_incoming', 'get_name', 'get_outgoing', 'get_predicates', 'get_predicates_for', 'get_tv', 'get_type', 'include_incoming', 'include_outgoing', 'is_source', 'is_valid', 'print_list', 'remove', 'set_av', 'set_tv', 'size', 'xget_atoms_by_av', 'xget_atoms_by_name', 'xget_atoms_by_target_atom', 'xget_atoms_by_type', 'xget_atoms_in_attentional_focus', 'xget_incoming', 'xget_outgoing', 'xget_predicates', 'xget_predicates_for']```

When I test the samething in early of April version,

Result of ```print atomspace``` in python agent:
```<opencog.atomspace.AtomSpace object at 0x7f6c8cdb9390>```

Result of ```print AtomSpace()``` in cogserver python shell:
```<opencog.atomspace.AtomSpace object at 0x7f6c8cdb9390>```

Result of ```print dir(atomspace)``` = Result of ```print dir(AtomSpace())```
```['__class__', '__contains__', '__delattr__', '__doc__', '__format__', '__getattribute__', '__getitem__', '__hash__', '__init__', '__iter__', '__len__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'add', 'add_link', 'add_node', 'clear', 'get_atom_string', 'get_atoms_by_av', 'get_atoms_by_name', 'get_atoms_by_target_atom', 'get_atoms_by_type', 'get_atoms_in_attentional_focus', 'get_av', 'get_incoming', 'get_name', 'get_outgoing', 'get_tv', 'get_type', 'include_incoming', 'include_outgoing', 'is_source', 'is_valid', 'print_list', 'remove', 'set_av', 'set_tv', 'size', 'xget_atoms_by_av', 'xget_atoms_by_name', 'xget_atoms_by_target_atom', 'xget_atoms_by_type', 'xget_atoms_in_attentional_focus', 'xget_incoming', 'xget_outgoing']```

**Comment by kim135797531** [comment]
I found temporary solution only works for cogserver.

Change this

```python
class ConceptualBlendingAgent(opencog.cogserver.MindAgent):
    def run(self, atomspace):
        atomspace.add(types.Node, "test")
```
to this
```python
class ConceptualBlendingAgent(opencog.cogserver.MindAgent):
    def run(self, atomspace):
        a = get_server_atomspace()
        initialize_opencog(a)
        a.add(types.Node, "test")
```

Yes, This solution doesn't use a given atomspace instance but request to server. It seems a little strange but it works well... I'll try to find the better solution.

Here is more debug information.
```python
print "get_server_atomspace(): " + str(get_server_atomspace())
print "atomspace: " + str(atomspace)
print "python shell: " + str(opencog.atomspace.AtomSpace())
```

Result:
```
get_server_atomspace(): <opencog.atomspace.AtomSpace object at 0x7fc486242390>
atomspace: <opencog.agent_finder.AtomSpace object at 0x7fc4862423d0>
python shell: <opencog.atomspace.AtomSpace object at 0x7fc4862423f0>
```

**Comment by linas** [comment]
I think your fix is correct.  Agents should almost always use the main cogserver atomsapce, instead of some other atomspace.  The fact that the run() method even takes an atomspace argument is a sign of some kind of weird design broken-ness.  I suggest just completely removing the atomspace argument from run(). 

Err, I think @cosmo might understand the python agent design, perhaps he can comment?  Is there ever any valid use-case for giving an agen some atomspace, other than the cogservers atomspace?

Ditto for the python shell: it makes no sense at all, to me, to have the python shell use some other atomspace (a third one! different from the first two!), so that looks like another bug.

**Comment by Cosmo** [comment]
his name is @cosmoharrigan, @linas.
i'm always getting notifications from you guys :)

**Comment by cosmoharrigan** [comment]
As a simple answer, I would say: no. The agent should just work with the default atomspace.

Perhaps there are hypothetical cases in which that is not the case, but to keep it simple for now (in the sense of not engineering things that aren't actually driven by user need), I think that is a good assumption.

I'll also include a link here to the thread where I suggested discussing a replacement design:
https://groups.google.com/d/msg/opencog/3iwRfYtx6hA/KW4mzMcO_m0J

**Comment by kim135797531** [comment]
Closed #1632
